q2_moshpit/eggnog/_dbs.py

```diff
@@ -174,3 +177,141 @@ def fetch_eggnog_proteins() -> EggnogProteinSequencesDirFmt:
+    source_path = os.path.join(str(eggnog_proteins), "ref_db.dmnd")
+    destination_path = os.path.join(str(diamond_db), "ref_db.dmnd")
+    shutil.move(source_path, destination_path)
+
+    # Return objects
+    return diamond_db
+
+
+def _validate_taxon_id(eggnog_proteins, taxon):
+    # Validate taxon id number
+    # Read in valid taxon ids
+    taxid_info = pd.read_csv(
+        os.path.join(str(eggnog_proteins), "e5.taxid_info.tsv"),
+        sep="\t"
+    )
+
+    # Convert them into a set
+    tax_ids = set()
+    for lineage in taxid_info["Taxid Lineage"]:
+        tax_ids.update(
+            set(
+                lineage.strip().split(",")
+            )
+        )
+
+    # Check for overlap with provided taxon id
+    if not tax_ids.intersection(set(str(taxon))):
+        raise ValueError(
+            f"'{taxon}' is not valid taxon ID. "
+            "To view all valid taxon IDs inspect e5.taxid_info.tsv "
+            "file in the input eggnog_proteins input."
+        )
+
+
+def fetch_ncbi_taxonomy() -> NCBITaxonomyDirFmt:
+    """
+    Script fetches 3 files from the internet and puts them into the folder of
+    a NCBITaxonomyDirFmt object.
+    """
+    # Initialize output object and paths
+    ncbi_data = NCBITaxonomyDirFmt()
+    zip_path = os.path.join(str(ncbi_data), "taxdmp.zip")
+    nodes_path = os.path.join(str(ncbi_data), "nodes.dmp")
+    names_path = os.path.join(str(ncbi_data), "names.dmp")
+    proteins_path = os.path.join(str(ncbi_data), "prot.accession2taxid.gz")
+    version_path = os.path.join(str(ncbi_data), "version.tsv")
+
+    # Download zip file
+    print(colorify("Downloading *.dmp files..."))
+    run_command(
+        cmd=[
+            "wget", "-O", zip_path,
+            "ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip"
+        ]
+    )
+
+    # Unzip
+    run_command(
+        cmd=[
+            "unzip", "-j", zip_path, "names.dmp", "nodes.dmp",
+            "-d", str(ncbi_data)
+        ]
+    )
+
+    # Remove zip file
+    run_command(cmd=["rm", zip_path])
+
+    # Download proteins
+    print(colorify("Downloading proteins file (~15 GB)..."))
+    run_command(
+        cmd=[
+            "wget", "-O", proteins_path,
+            "ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/"
+            "prot.accession2taxid.gz"
+        ]
+    )
+
+    # Constructing version file
+    print(colorify("Constructing version file..."))
+    _write_version_tsv(nodes_path, names_path, proteins_path, version_path)
+
+    # Return object
+    print(colorify(
+        "Done! Moving data from temporary directory to final location..."
+    ))
+    return ncbi_data
+
+
+def _write_version_tsv(nodes, names, proteins, version):
+    names_time = datetime.date.fromtimestamp(os.path.getmtime(nodes))
+    nodes_time = datetime.date.fromtimestamp(os.path.getmtime(names))
+    proteins_time = datetime.date.fromtimestamp(os.path.getmtime(proteins))
```

Comment:
The issue I have with this approach is that it says that the version of the files I fetched today is today midnight - I find it hard to believe it though... What date/time did you get when you tried it out?

Reply:
I think they are updated every day. I also found it weird. But if you look at the other files they don't get updated as often so it's not like they are all getting a `touch` every day. 

By running `wget --spider --server-response ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip` one gets

```
lr--r--r--   1 ftp      anonymous       35 Oct 27  2017 Ccode_dump.txt -> biocollections/Collection_codes.txt
lr--r--r--   1 ftp      anonymous       36 Oct 27  2017 Cowner_dump.txt -> biocollections/Institution_codes.txt
lr--r--r--   1 ftp      anonymous       43 Oct 27  2017 Icode_dump.txt -> biocollections/Unique_institution_codes.txt
-r--r--r--   1 ftp      anonymous     2774 Apr 13  2023 Major_taxonomic_updates_2023.txt
dr-xr-xr-x   2 ftp      anonymous    12288 Jan 18 20:09 accession2taxid
dr-xr-xr-x   2 ftp      anonymous     4096 Apr 28  2022 biocollections
-r--r--r--   1 ftp      anonymous   495372 Jan 19 10:12 coll_dump.txt
-r--r--r--   1 ftp      anonymous     1715 May 15  2018 ncbi_taxonomy_genussp.txt
dr-xr-xr-x   3 ftp      anonymous     4096 Jan 19 15:42 new_taxdump
-r--r--r--   1 ftp      anonymous 13582413 Jan 19 15:37 taxcat.tar.Z
-r--r--r--   1 ftp      anonymous       47 Jan 19 15:37 taxcat.tar.Z.md5
-r--r--r--   1 ftp      anonymous 10045614 Jan 19 15:37 taxcat.tar.gz
-r--r--r--   1 ftp      anonymous       48 Jan 19 15:37 taxcat.tar.gz.md5
-r--r--r--   1 ftp      anonymous 10045475 Jan 19 15:37 taxcat.zip
-r--r--r--   1 ftp      anonymous       45 Jan 19 15:37 taxcat.zip.md5
-r--r--r--   1 ftp      anonymous      655 Jun 29  2016 taxcat_readme.txt
-r--r--r--   1 ftp      anonymous 64176832 Jan 19 15:29 taxdmp.zip
-r--r--r--   1 ftp      anonymous       45 Jan 19 15:29 taxdmp.zip.md5
-r--r--r--   1 ftp      anonymous 122832711 Jan 19 15:29 taxdump.tar.Z
-r--r--r--   1 ftp      anonymous       48 Jan 19 15:29 taxdump.tar.Z.md5
-r--r--r--   1 ftp      anonymous 64178144 Jan 19 15:29 taxdump.tar.gz
-r--r--r--   1 ftp      anonymous       49 Jan 19 15:29 taxdump.tar.gz.md5
dr-xr-xr-x   2 ftp      anonymous    12288 Jan  1 05:00 taxdump_archive
-r--r--r--   1 ftp      anonymous     5443 Apr 27  2023 taxdump_readme.txt
```

Reply:
OK, so I've looked into it a bit more and their documentation says that they update them every hour (!). Unfortunately, in this case, I don't really see why we would need to store those versions in a separate file - the provenance already does it for us so no need to complicate it more. However, there are two other things that occurred to me while reading through their docs:
- they distinguish a "newer" taxonomy URL and taxdump and suggest one should use that one - I can also see that the new taxdump files are included under the link you are using - maybe let's stick with using the same link but instead of fetching the `taxdmp.zip` we fetch the `new_taxdump/new_taxdump.zip`? Can you check what's different about its contents? (maybe it needs the semantic type update...)
- they provide MD5 hash files for all of the data - we should use those to check that the data you fetch is not corrupted - you can use the hash provided in those files to cross-check with the files you fetch to see whether their MD5 hash matches (and error out if it doesn't)
- this also implies that we should park the other PR for now - there is not much use for it now (let's not throw the code out - you can push your branch to the upstream, in case we need it in the future again for some other database)

Finally, here's the link to the documentation: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7408187/

Reply:
@misialq 
> they distinguish a "newer" taxonomy URL and taxdump and suggest one should use that one - I can also see that the new taxdump files are included under the link you are using - maybe let's stick with using the same link but instead of fetching the taxdmp.zip we fetch the new_taxdump/new_taxdump.zip? Can you check what's different about its contents? (maybe it needs the semantic type update...)

- `new_taxdump.zip` also has `nodes.dmp` and `names.dmp` and passes the validations that you wrote for those type of files. But I dont know if diamond would break with the new_taxdump files. They link the other files in their tutorial. 

> they provide MD5 hash files for all of the data - we should use those to check that the data you fetch is not corrupted - you can use the hash provided in those files to cross-check with the files you fetch to see whether their MD5 hash matches (and error out if it doesn't)

- I'll add the check. Should I add it inside `q2_moshpit/_utils.py`? I think this might be useful in other modules, including what I working on atm for BUSCO. They also provide md5 files. 

> this also implies that we should park the other PR for now - there is not much use for it now (let's not throw the code out - you can push your branch to the upstream, in case we need it in the future again for some other database)

- Yeah I see, I agree. Should I close the PR then and push the branch to the upstream?

Reply:
> new_taxdump.zip also has nodes.dmp and names.dmp and passes the validations that you wrote for those type of files. But I dont know if diamond would break with the new_taxdump files. [They link the other files in their tutorial.](https://github.com/bbuchfink/diamond/wiki/3.-Command-line-options#:~:text=current%20working%20directory.-,Makedb%20options,ftp%3A//ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip.,-General%20options)

OK, let's do it this way: let's keep using the one they have, since we know it works, and open an issue here to investigate swapping for the newer files. Could you please take care of that?

>Should I add it inside q2_moshpit/_utils.py?

Yeah, I think that would be a good place for it. 👌 

>Should I close the PR then and push the branch to the upstream?

Yes, please, thanks! 🙏